### PR TITLE
module: add a gpufreq module for interacting with GPUs.

### PR DESCRIPTION
**Pull request by kdub**
Add a module for controlling and interacting with GPUs. The module
currently supports kgsl/adreno based GPUs. This allows for querying
and setting the governor on the GPU, as well as seeing the frequencies
at which the the GPU can can operate.

Change-Id: I02bb997b51426ddaa13e1f8da375aa4c4a0d341a

### CODE REVIEW

File: devlib/module/gpufreq.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+from devlib.exception import TargetError
+
+class GpufreqModule(Module):
+
+    name = 'gpufreq'
+    path = ''
+
+    def __init__(self, target):
+        super(GpufreqModule, self).__init__(target)
+        frequencies_str = self.target.read_value("sys/kernel/gpu/gpu_freq_table")
```
Comment (setrofim) [inline devlib/module/gpufreq.py]: Missing initial `/` in path.

File: devlib/module/gpufreq.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        self.frequencies = map(int, frequencies_str.split(" "))
+        self.frequencies.sort()
+        self.governors = self.target.read_value("sys/kernel/gpu/gpu_available_governor").split(" ")
```
Comment (setrofim) [inline devlib/module/gpufreq.py]: Missing initial `/` in path.

File: devlib/module/gpufreq.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    def probe(target):
+        # kgsl/Adreno
+        probe_path = '/sys/kernel/gpu/'
+        if target.file_exists(probe_path):
+            model = target.read_value(probe_path + "gpu_model")
+            if re.search('adreno', model, re.IGNORECASE):
+                return True
+        return False
+
+    def get_governors(self):
```
Comment (setrofim) [inline devlib/module/gpufreq.py]: This method is not necessary. Calling code should just use `self.governors` directly.

File: devlib/module/gpufreq.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        Warning, this method does not check if the gpu is online or not. It will
+        try to read the current frequency and the following exception will be
+        raised ::
+
+        :raises: TargetError if for some reason the frequency could not be read.
+
+        """
+        return int(self.target.read_value("sys/kernel/gpu/gpu_clock"))
+
+    def get_model_name(self):
```
Comment (setrofim) [inline devlib/module/gpufreq.py]: Should add `@memoized` as this will not change over the course of a run.

File: devlib/module/gpufreq.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    def get_governors(self):
+        """
+        Returns the list of available governors for the GPU
+        """
+        return self.governors
+
+    def set_governor(self, governor):
+        if governor not in self.governors:
+            raise TargetError('Governor {} not supported for gpu {}'.format(governor, cpu))
+        self.target.write_value("sys/kernel/gpu/gpu_governor", governor)
```
Comment (setrofim) [inline devlib/module/gpufreq.py]: Missing initial `/` in path.

**Comment by kdub** [comment]
@setrofim Thank you for the review! Corrected the problems.

**Comment by setrofim** [comment]
`memoized` needs to be imported from `devlib.utils.misc`. Other than that, looks good.

**Comment by kdub** [comment]
@setrofim thanks corrected again

**Pull request merged**